File: 1248. Count Number of Nice Subarrays/numberOfSubarrays.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const numberOfSubarrays = (nums, k) => {
  let left = 0;
  let middle = 0;
  let oddCnt = 0;
  let res = 0;

  for (let r = 0; r < nums.length; r++) {
    if (nums[r] % 2) {
      oddCnt++;
    }

    while (oddCnt > k) {
      if (nums[left] % 2) {
        oddCnt--;
      }
      left++;
      middle = left;
    }

    if (oddCnt === k) {
      while (nums[middle] % 2 === 0) {
        middle++;
      }
      res += (middle - left) + 1;
    }
  }

  return res;
};


const nums = [1, 1, 2, 1, 1];
const k = 3;
console.log('numberOfSubarrays: ', numberOfSubarrays(nums, k));

File: 1248. Count Number of Nice Subarrays/numberOfSubarrays.py
from typing import List

class Solution:
  def numberOfSubarrays(self, nums: List[int], k: int) -> int:
    left = 0
    middle = 0
    odd_cnt = 0
    res = 0
  
    for right in range(len(nums)):
      if nums[right] % 2:
        odd_cnt += 1
      
      while odd_cnt > k:
        if nums[left] % 2:
          odd_cnt -= 1
        left += 1
        middle = left
        
      if odd_cnt == k:
        while not nums[middle] % 2:
          middle += 1
        res += (middle - left) + 1
      
    return res

 
nums = [1, 1, 2, 1, 1]
k = 3
solution = Solution()
print('numberOfSubarrays: ', solution.numberOfSubarrays(nums, k))
